Code/GraphMol/MolDraw2D/DrawMolMCHLasso.cpp

```diff
@@ -172,35 +171,321 @@ double getLassoWidth(const DrawMolMCH *dm, int atNum, int lassoNum) {
+// going anti-clockwise. Also returns the angle for where the line
+// between the centres intersects with the circle on at1Cds (bang).
+void calcIntersectingArcAngles(const Point2D &at1Cds, double rad1,
+                               const Point2D &at2Cds, double rad2, double &ang1,
+                               double &ang2, double &bang) {
+  Point2D isect1, isect2;
+  intersectingCirclePoints(at1Cds, rad1, at2Cds, rad2, isect1, isect2);
+  bool swapped;
+  calcSubtendedAngles(isect1, isect2, at1Cds, at2Cds, ang1, ang2, bang,
+                      swapped);
+}
+
+// See if the arc defined by the centre, radius etc. has either of its ends
+// inside a circle of one of the otherAtoms
+bool areArcEndsInOtherCircle(const Point2D &centre, double radius, double ang1,
+                             double ang2, size_t atNum, const DrawMolMCH *dm,
+                             const std::vector<unsigned int> &otherAtoms,
+                             int lassoNum) {
+  ang1 *= M_PI / 180.0;
+  Point2D end1{centre.x + radius * cos(ang1), centre.y + radius * sin(ang1)};
+  ang2 *= M_PI / 180.0;
+  Point2D end2{centre.x + radius * cos(ang2), centre.y + radius * sin(ang2)};
+
+  for (size_t i = 0; i < otherAtoms.size(); ++i) {
+    if (otherAtoms[i] == atNum) {
       continue;
     }
-    double lassoWidth = getLassoWidth(this, ca, lassoNum);
-    Point2D radii(lassoWidth, lassoWidth);
-    std::vector<Point2D> pts{atCds_[ca], radii};
-    DrawShapeArc *ell = new DrawShapeArc(pts, 0.0, 360.0, lineWidth,
-                                         scaleLineWidth, col, false, ca);
-    arcs.emplace_back(ell);
+    auto otherRad = getLassoWidth(dm, otherAtoms[i], lassoNum);
+    if ((end1 - dm->atCds_[otherAtoms[i]]).lengthSq() - otherRad * otherRad <
+        -1.0e-4) {
+      return true;
+    }
+    if ((end2 - dm->atCds_[otherAtoms[i]]).lengthSq() - otherRad * otherRad <
+        -1.0e-4) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Find the start of any existing arcs involving atom i, feeding them into
+// arcAngs and bangs so as to be able to merge in any new arcs that intersect
+// them.  Removes the existing arcs if found.
+void addExistingArcs(size_t i,
+                     std::vector<std::unique_ptr<DrawShapeArc>> &currArcs,
+                     std::vector<double> &arcAngs,
+                     std::vector<std::pair<double, unsigned int>> &bangs) {
+  for (auto &arc : currArcs) {
+    if (arc->atom1_ == static_cast<int>(i)) {
+      arcAngs.push_back(arc->ang2_);
+      arcAngs.push_back(arc->ang1_);
+      if (arc->ang2_ < arc->ang1_) {
+        bangs.push_back(
+            std::make_pair((arc->ang2_ + arc->ang1_) / 2.0, bangs.size()));
+      } else {
+        double bang = -360.0 + (arc->ang2_ + arc->ang1_ + 360.0) / 2.0;
+        bangs.push_back(std::make_pair(bang, bangs.size()));
+      }
+      arc.reset();
+    }
+  }
+  currArcs.erase(
+      remove_if(currArcs.begin(), currArcs.end(),
```

Comment:
This probably should be
```suggestion
      std::remove_if(currArcs.begin(), currArcs.end(),
```

Reply:
Thanks.  I have pushed that and also a gratuitous grammar fix in one of the comments.
Let's hope it all works better this time.